### ISSUE TITLE
[HOLD for payment 2023-04-12] [$2000] City and State doesn’t populate for the address I-80 but works well with other address like I-35 , I-40, I-70

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#15753]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open chrome staging dot
2. Go to personal information and address
3. In the address field type in : I
4. Try to select I-80 and see that the city and state doesn't populate with the recommended address
5. But if you select I-35, I-40 or I-70, every recommended address like city , state, and country populates without a problem

## Expected Result:
Setting address I-80 (Interstate 80) should populate the addresses as per the recommendation

## Actual Result:
Setting address I-80 (Interstate 80) doesn’t populate the addresses as per the recommendation

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.80-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/223845024-748c7542-5ac6-4403-b7e8-8456ce2832a5.mp4

https://user-images.githubusercontent.com/43996225/223845043-fdd65945-6548-400b-9350-7cd817bb8345.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @priya-zha
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1678254121686449

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0195f9f8aaab8dc5c1</li>
        <li>Upwork Job ID: 1634338652687765504</li>
        <li>Last Price Increase: 2023-03-17</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @arielgreen (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by arielgreen** [comment]
Note that even for other ones (e.g., I-35), it doesn't populate zip.

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0195f9f8aaab8dc5c1

**Comment by MelvinBot** [comment]
Current assignee @arielgreen is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @sobitneupane (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @francoisl (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by tienifr** [comment]
Possibly due to Google Map API result. The code in https://github.com/Expensify/App/blob/a34c4a7da4ea7db905c28fff1ebb46850c2affe3/src/components/AddressSearch.js#L194 will fetch the location detail from `/api?command=Proxy_GooglePlaces`, and update the city, state, country values in the input.

For address such as I-36, the location api returns city and state info
<img width="1063" alt="image" src="https://user-images.githubusercontent.com/113963320/224462139-c2ccc3d1-1de5-4633-b4ac-047bf49869c5.png">


For address such us I-80, the location api **does not return city and state** info
<img width="1068" alt="image" src="https://user-images.githubusercontent.com/113963320/224462105-0b1fb574-8b85-40e5-8dc9-f3d398a07f7d.png">

**Comment by sshere** [comment]
Contributor details
Your Expensify account email: sshere1985@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~0116ef5450a33cc027

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Setting address I-80 (Interstate 80) doesn’t populate the addresses (city, state, ...) as per the Google Places Autocomplete

### What is the root cause of that problem?
The root cause is:
1. When we fetch the data from the Google Places Autocomplete, it tries to give us predictions on the places even though the places might not exist in Google's database
2. When clicking on the place in the autocomplete list, we retrieve the place details from the Google Place API, which will only return the address that exists Google's database
3. The autocomplete option for 'I-80' query does not exist in Google's database, so when we get its details, Google will try to find the closest address in its database and return it, in this case the address in Google's database does not have information like city, state, ...

For more information please refer to Google's response here: https://issuetracker.google.com/issues/35823492#comment2
This is a known behavior.

### What changes do you think we should make in order to solve the problem?

To fix this, we need to use the Google's place autocomplete information as a fallback in case the Place Detail does not have that information.
For example, when we query for `I 80` and click on the `I-80, Seward, NE, USA` result in the autocomplete dropdown.
If we don't have the state from the Place Detail API, we'll use `NE` from the autocomplete result as a fallback.
If we don't have the city from the Place Detail API, we'll use `Seward` from the autocomplete result as a fallback.

The response from Google Autocomplete is quite structured so we can retrieve the state and city using the following logic:
```
const fallbackData = autocompleteData.terms;
const stateAutocompleteFallback = fallbackData.length >= 2 ? fallbackData[fallbackData.length - 2].value : '';
const cityAutocompleteFallback = fallbackData.length >= 3 ? fallbackData[fallbackData.length - 3].value : '';
```
in https://github.com/Expensify/App/blob/4a172c9e9a2bcf02b4a3b0cec3537fdf82195d73/src/components/AddressSearch.js#L85

The `autocompleteData` can be passed into `saveLocationDetails` by using the `data` returned in https://github.com/Expensify/App/blob/4a172c9e9a2bcf02b4a3b0cec3537fdf82195d73/src/components/AddressSearch.js#L157.

There might we wild result format returned from Google Autocomplete for countries other than the US, if we're not sure we can limit the above fallback logic to addresses in the US. It will improve the user experience.

### What alternative solutions did you explore? (Optional)
NA


### Result
Working well after the fix:

https://user-images.githubusercontent.com/113963320/224467227-5f016b7d-02bf-4222-a2db-724673c45220.mov

**Comment by sobitneupane** [comment]
Thanks for the [proposal](https://github.com/Expensify/App/issues/15753#issuecomment-1464833715) @tienifr. 

But I don't think we can go with your proposal because it depends on the structure of data returned from an api which is not always consistent.

**Comment by tienifr** [comment]
@sobitneupane sure, the structure of the address within the US is very consistent as I observed.

I agree my solution will be the last one to be considered if there's no other perfect solution.

**Comment by francoisl** [comment]
I'd be ok with this solution, though it sounds like it wouldn't work very well for all non-US addresses. Looking at the `terms` array for random addresses from other countries, the city is not always at index `fallbackData.length - 3`, and the state is not always at index `fallbackData.length - 2`, because other countries don't necessarily use the same `<street number> <street name> <city> <state>` address format.

Alternatively, we could use this fallback solution only if the predicted country is `USA`?

**Comment by tienifr** [comment]
> Alternatively, we could use this fallback solution only if the predicted country is USA?

@francoisl yes we can do this as well.

**Comment by sobitneupane** [comment]
@hoangzinh Will your proposal alone yield the expected output? Or should it be integrated alongside [this proposal](https://github.com/Expensify/App/issues/15753#issuecomment-1464833715).

Our expected output is 

> Setting address I-80 (Interstate 80) doesn’t populate the addresses as per the recommendation

The issue here is while selecting I-80 state, it doesn't populate 'City' , 'Sate' and 'zipcode'.
<img width="366" alt="Screenshot 2023-03-17 at 14 18 10" src="https://user-images.githubusercontent.com/25876548/225854170-a917e6d0-b551-4ea7-9795-94215ddeb373.png">

**Comment by tienifr** [comment]
@sobitneupane from [the screenshot of the proposal](https://github.com/Expensify/App/issues/15753#issuecomment-1472332636) it seems the address in the issue `I-80, Seward, NE, USA` disappears from the search result now, likely due to the changed query types. Not sure if we want that.

**Comment by francoisl** [comment]
> Set isLimitedToUSA={this.state.isUsaForm} instead of always false in component [AddressPage](https://github.com/Expensify/App/blob/c40148b5ddd82b54ddb40f7437d1bee22fbf5682/src/pages/settings/Profile/PersonalDetails/AddressPage.js#L72) => we can query by 'country:us` => more correct result

That means that the user would first need to manually change the country in the dropdown, right? It sounds like it would partially defeat the purpose of having the auto-complete search field. 


I agree there's no perfect solution here, but we can try to improve the auto-complete whenever possible instead. With that said, I think I'm leaning towards @tienifr's original [proposal](https://github.com/Expensify/App/issues/15753#issuecomment-1464833715) – _without_ limiting the fallback solution if the predicted country is `USA` as I [suggested](https://github.com/Expensify/App/issues/15753#issuecomment-1468974524). It will help populate the city and state for all countries that use the same `autocompleteData.terms` structure as the US, and for other countries it will remain the same as today.

Does that work with you @sobitneupane?

**Comment by arielgreen** [comment]
bump @sobitneupane

**Comment by sobitneupane** [comment]
I agree @francoisl. We can go with @tienifr's [proposal](https://github.com/Expensify/App/issues/15753#issuecomment-1464833715). 
My only concern is since the solution depends on the structure of response from an third-party api, it might break down if the api response changes.

**Comment by MelvinBot** [comment]
📣 @tienifr You have been assigned to this job by @arielgreen!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by tienifr** [comment]
@francoisl @sobitneupane @arielgreen The [PR](https://github.com/Expensify/App/pull/16388) is ready to review!

**Comment by tienifr** [comment]
# Regression Test Proposal

**Bug**: Typing emoji in CAPS doesn’t highlight the text on the emoji suggestion list 

**Proposed Test Steps**:

1. Login with any account
2. Click on Profile avatar > Profile > Personal details > Home address
3. In the Address line 1 field type in: **I**
4. Select **I-35** (i.e. I-35 Ottawa, KS, USA)
5. Verify that other fields are filled automatically just as the suggestion (i.e. Interstate 35 - Ottawa - Kansas - United States)
6. Select **I-80** (i.e. I-80 Seward, NE, USA)
7. Verify that other fields are filled automatically just as the suggestion (i.e. Interstate 80 - Seward - Nebraska - United States)
8. In the Address line 1 field, type in: N12 ireland
9. Select the option N12, Knockaconny, County Monaghan, Ireland
10. Make sure that the City, State / Province, and Country fields get automatically filled
11. In the Address line 1 field, type in: BR-101
12. Select the option that starts with BR-101, Linhares
13. Make sure that the City, State / Province, and Country fields get automatically filled

![image](https://user-images.githubusercontent.com/113963320/229409548-c26ed185-26a9-4337-9ed2-e2f1408cdb0a.png)

![image](https://user-images.githubusercontent.com/113963320/229409616-c31d1ea0-8863-49ef-a628-37d1255fd077.png)

![image](https://user-images.githubusercontent.com/113963320/229409740-ad37c0a1-cdfd-4c6a-9c58-5d35e3bbc992.png)

![image](https://user-images.githubusercontent.com/113963320/229409778-70a8ac0f-562d-4cf5-bad6-87b020c2fce5.png)

Do we :+1: or :-1:

**Comment by francoisl** [comment]
> Bug: Typing emoji in CAPS doesn’t highlight the text on the emoji suggestion list

This sounds like an accidental copy-paste from another issue? The rest of the steps sounds good to me.

**Comment by tienifr** [comment]
@francoisl Sorry, my bad, I just updated

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.94-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/16388


If no regressions arise, payment will be issued on 2023-04-12. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@sobitneupane / @francoisl] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@sobitneupane / @francoisl] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@sobitneupane / @francoisl] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@arielgreen] Determine if we should create a regression test for this bug.
- [ ] [@sobitneupane] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@arielgreen] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by francoisl** [comment]
I filled out most of the checklist above. For the regression test steps, we can use [this comment](https://github.com/Expensify/App/issues/15753#issuecomment-1493624702) @arielgreen.

**Comment by arielgreen** [comment]
@tienifr @sobitneupane @priya-zha, please apply for [this posting on Upwork](https://www.upwork.com/jobs/~016301285c775e4278) asap so I can send payment

**Comment by priya-zha** [comment]
@arielgreen Submitted the proposal. Thanks

**Comment by sobitneupane** [comment]
@arielgreen Proposal Submitted.

**Comment by francoisl** [comment]
@arielgreen friendly bump ^

**Comment by arielgreen** [comment]
@priya-zha @tienifr paid, apologies for delay

**Comment by arielgreen** [comment]
Just waiting on @sobitneupane to accept, then everything here is done and we can close.

**Comment by sobitneupane** [comment]
@arielgreen Accepted the offer.

**Comment by arielgreen** [comment]
All payments made, we are set here.
